### ISSUE TITLE
Use information from Github for Build Server integration

### ISSUE THREAD

**Issue by carl-menezes-aql** [gitextensions/gitextensions#8461]
## Enable a whole range of build systems by using Github as a placeholder

Right now, if you want to see the build status of your commits, you need to enable (and configure) a specific build server integration. There are two problems with this:

1. There are very few available
2. Adding support for the majority out there is a losing battle

![image](https://user-images.githubusercontent.com/28543618/93408536-eec1c980-f8e8-11ea-81bc-e75e3f14453e.png)

However, online services like Github, GitLab, etc. already have this information - in order to verify PRs/merge requests, build systems already push build statuses to these services.
![image](https://user-images.githubusercontent.com/28543618/93408825-9a6b1980-f8e9-11ea-922d-ceb6009ead60.png)

Therefore, if this information can be displayed within Git Extensions itself, chances are that it would satisfy a lot of requests for various build systems. Further, configuration would be simple: you would most likely only need an API token.

## Environment

- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4220.0
- DPI 96dpi (no scaling)

**Comment by RussKie** [comment]
How would we do this?

**Comment by carl-menezes-aql** [comment]
With Gihub, you could use the GraphQL statusCheckRollup query. There is also a REST API version, but that is subject to quota limits. 

I haven't used the GitLab API much, but I daresay that even implementing it for GitHub initially would deliver a heap of value.

Here's an example for github.com/drone/drone:
```
{
  repository(name: "drone", owner: "drone") {
    ref(qualifiedName: "master") {
      target {
        ... on Commit {
          history(first: 5) {
            nodes{
              statusCheckRollup{
                commit{
                  oid
                  commitUrl
                }
                state
              }
            }
          }
        }
      }
    }
  }
}
```

Here is the result:
```
{
  "data": {
    "repository": {
      "ref": {
        "target": {
          "history": {
            "nodes": [
              {
                "statusCheckRollup": {
                  "commit": {
                    "oid": "1cdeb5f84d0b5f25265588ad11efaebe6f6b5af0",
                    "commitUrl": "https://github.com/drone/drone/commit/1cdeb5f84d0b5f25265588ad11efaebe6f6b5af0"
                  },
                  "state": "SUCCESS"
                }
              },
              {
                "statusCheckRollup": null
              },
              {
                "statusCheckRollup": null
              },
              {
                "statusCheckRollup": {
                  "commit": {
                    "oid": "2565aa31798ac8e1dcfe2cdddf976c670a689e61",
                    "commitUrl": "https://github.com/drone/drone/commit/2565aa31798ac8e1dcfe2cdddf976c670a689e61"
                  },
                  "state": "SUCCESS"
                }
              },
              {
                "statusCheckRollup": {
                  "commit": {
                    "oid": "677d7cc2f75d36541d24c12b74df6b516001af67",
                    "commitUrl": "https://github.com/drone/drone/commit/677d7cc2f75d36541d24c12b74df6b516001af67"
                  },
                  "state": "SUCCESS"
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

**Comment by gerhardol** [comment]
But this requires that the buildserver updates the Git server, normally only for PRs.
Builds can be started on for instance branches, without updating the Git server.

BitBucket Server documentation
https://docs.atlassian.com/bitbucket-server/rest/7.6.0/bitbucket-build-rest.html
No info about tests it seem

This could simplify for GitHub and is an alternative, but there are quite some Git servers too, no universal solution.

**Comment by carl-menezes-aql** [comment]
Yes, it does require the build server to update the Git server. However, with most software development, that is what happens already. 

Further, whether it is per commit or per PR would depend on the development model in use. When forks are used, then the merge commit (and maybe head) would typically have a build status. However, if a branch is used, then more commits would have a build status. The point is that the information already exists - it would be a matter of querying and displaying it.

I agree that there is no universal solution. However, given how widely Github is used and how many build systems integrate with it, it would mean that Git Extensions could display commit build status:

1. When the build server is not reachable
2. For a much larger selection of build systems than it currently does

The same logic above could then be used to integrate other git providers as needed and users would benefit from commit statuses in their local client.
